adr/0009-extension-structure.adoc

```diff
@@ -0,0 +1,261 @@
+An example of ambiguity are CDI events: they can be used by application code (API) but also allow other extensions to contribute (SPI).
+
+To avoid this, let's only consider APIs and merge API and SPI into a single `api` module, potentially with subpackages for different concerns (e.g., `io.quarkus.extension.api.spi`).
+
+==== The Coupling Dimension
+
+An extension using another extension can be either strongly or loosely coupled, depending on whether it requires the other extension to be present at runtime:
+
+* Strongly coupled: An extension directly depends on another extension's runtime module (and its deployment to the deployment module), requiring it to be present.
+* Loosely coupled: An extension only depends on the public API or SPI of another extension, allowing it to be used without requiring the full extension.
+
+When an extension depends on another extension, it should clearly indicate whether it is tightly or loosely coupled:
+
+* Tightly coupled: The extension depends on the `runtime` or `deployment` module of another extension, indicating that it requires the full extension to be present.
+* Loosely coupled: The extension depends on the `api` or `deployment-api` module of another extension, indicating that it can work independently of the full extension. Note that this may require conditional logic to handle the absence of the extension at runtime.
+
+An extension can also decide to only support _tight-coupling_ and does not provide a `runtime-api` or `deployment-api` module.
+
+[#the-place-of-the-public-api]
+==== The place of the public API
+
+Historically, extensions have often placed their public APIs in the `runtime` module.This forces a tight coupling between the extension and its consumers, as they must depend on the `runtime` module to access the public API.
+To circumvent this, a `runtime-spi` module was introduced, initially limited to _SPIs_, but later expanded to include public APIs as well.
+
+This has led to confusion about the intended use of these modules and their separation.
+To clarify this, we propose the following rules:
+
+* Public APIs should be placed in a dedicated `runtime-api` module, which can be used independently of the full extension.
+* The `runtime` module should focus on the internal implementation and runtime logic, delegating public APIs to the `api` module.
+* The `runtime-api` module contains both the public API and the SPI, allowing other extensions to depend on it without pulling in the full extension.
+
+For extensions requiring tight-coupling, the `runtime` module can still be used to expose public APIs, but this should be avoided when possible.
+
+[cols="1,3,2",options="header"]
+|===
+| Scenario | Recommended Structure | Notes
+
+| Small/simple extension not reused by others
+| `runtime`
+| Keep everything in `runtime`. Avoid unnecessary modularity.
+
+| Extension exposes public types used by application code or other extensions
```

Comment:
I think most extensions have that.

Reply:
In my opinion, it should be the only mandatory module. But that’s not the case. Typically, the dev service does not have one (yet), and JMS-SPI does not either (but that extension is a bad example).